eth/downloader/downloader.go

```diff
@@ -764,12 +775,51 @@ func calculateRequestSpan(remoteHeight, localHeight uint64) (int64, int, int, ui
 	return int64(from), count, span - 1, uint64(max)
 }
 
+// isValidChain checks if the chain we're about to receive from this peer is valid or not
+// in terms of reorgs. We won't reorg beyond the last bor checkpoint submitted to mainchain.
+func (d *Downloader) isValidChain(p *peerConnection, remoteHeader *types.Header) (bool, error) {
+	// We want to validate the chain by comparing the last checkpointed block
+	// we're storing in `checkpointWhitelist` with the peer's block.
+
+	// Check for availaibility of the last checkpointed block.
+	// This can be also be empty if our heimdall is not responsing
+	// or we're running without it.
+	if len(d.checkpointWhitelist) <= 0 {
+		// worst case, we don't have the checkpoints in memory
+		return true, nil
```

Comment:
shall we trust then to a remote peer?

Reply:
I am not really sure. It's just that there can be multiple reasons for not having an entry in checkpoint whitelist map (node restart, heimdall down/syncing, running without heimdall, etc).

Reply:
I believe that by the end of the day we only trust to Heimdall node. Let's discuss such cases on the next reorg call.